src/cursor/cur_std.c

```diff
@@ -556,6 +569,180 @@ err:		cursor->saved_err = ret;
 	API_END(session, ret);
 }
 
+/*
+ * __wt_cursor_cache --
+ *	Add this cursor to the cache.
+ */
+int
+__wt_cursor_cache(WT_CURSOR *cursor, WT_DATA_HANDLE *dhandle)
+{
+	WT_SESSION_IMPL *session;
+
+	session = (WT_SESSION_IMPL *)cursor->session;
+	WT_ASSERT(session, !F_ISSET(cursor, WT_CURSTD_CACHED) &&
+	    (dhandle != NULL || F_ISSET(cursor, WT_CURSTD_CACHE_CHILD)));
+
+	if (!F_ISSET(cursor, WT_CURSTD_CACHE_CHILD))
+		WT_RET(cursor->reset(cursor));
+	if (dhandle != NULL) {
+		WT_DHANDLE_CACHE(dhandle);
+
+		/*
+		 * Acquire a reference while decrementing the in-use counter.
+		 * After this point, the dhandle may be marked dead, but the
+		 * actual handle won't be removed.
+		 */
+		session->dhandle = dhandle;
+		WT_DHANDLE_ACQUIRE(dhandle);
+		__wt_cursor_dhandle_decr_use(session);
+	}
+	F_SET(cursor, WT_CURSTD_CACHED);
+	return (0);
+}
+
+/*
+ * __wt_cursor_reopen --
+ *	Reopen this cursor from the cached state.
+ */
+int
+__wt_cursor_reopen(WT_CURSOR *cursor, WT_DATA_HANDLE *dhandle)
+{
+	WT_DECL_RET;
+	WT_SESSION_IMPL *session;
+
+	session = (WT_SESSION_IMPL *)cursor->session;
+	WT_ASSERT(session, F_ISSET(cursor, WT_CURSTD_CACHED));
+
+	if (dhandle != NULL) {
+		/*
+		 * A WT_NOTFOUND return is a signal to the caller to close
+		 * the cursor. We still need to restore the cursor so the
+		 * close can function normally.
+		 */
+		if (WT_DHANDLE_INACTIVE(dhandle) ||
+		    !F_ISSET(dhandle, WT_DHANDLE_OPEN) ||
+		    F_ISSET(dhandle, WT_DHANDLE_DROPPED))
+			ret = WT_NOTFOUND;
+
+		session->dhandle = dhandle;
+		__wt_cursor_dhandle_incr_use(session);
+		WT_DHANDLE_RELEASE(dhandle);
+		WT_DHANDLE_UNCACHE(dhandle);
+	}
+	F_CLR(cursor, WT_CURSTD_CACHED);
+	return (ret);
+}
+
+/*
+ * __wt_cursor_close_cache --
+ *	Put the cursor into a cached state, called during close operations.
+ */
+int
+__wt_cursor_close_cache(WT_SESSION_IMPL *session, WT_CURSOR *cursor)
+{
+	WT_DECL_RET;
+
+	if ((ret = cursor->cache(cursor)) == 0) {
+		WT_STAT_CONN_INCR(session, cursor_cache);
+		WT_STAT_DATA_INCR(session, cursor_cache);
+		return (0);
+	}
+
+	/*
+	 * TODO: The caching operation failed, so we'll continue the close,
+	 * reopen first??
+	 */
+	WT_TRET(cursor->reopen(cursor));
+	return (ret);
+}
+
+/*
+ * __wt_cursor_open_cache --
+ *	Open a matching cursor from the cache.
+ */
+int
+__wt_cursor_open_cache(WT_SESSION_IMPL *session, const char *uri,
+    const char *cfg[], WT_CURSOR **cursorp)
+{
+	WT_CONFIG_ITEM cval;
+	WT_CURSOR *cursor;
+	WT_DECL_RET;
+	const char *tmp_cfg;
+
+	if (cfg == NULL) {
+		tmp_cfg = NULL;
+		cfg = &tmp_cfg;
+	}
+	WT_RET(__wt_config_gets(session, cfg, "readonly", &cval));
```

Comment:
I suggest using `__wt_config_gets_def` consistently (and checking the defaults), to avoid repeated parsing of config string defaults on every cursor (re)open.

Reply:
Yes. I found this to be a hot spot with profiling.